File: filesystem/src/test/scala/net/kemitix/thorp/filesystem/FileSystemTest.scala
package net.kemitix.thorp.filesystem

import scala.jdk.CollectionConverters._
import net.kemitix.thorp.domain.{RemoteKey, Sources}
import org.scalatest.FreeSpec
import zio.DefaultRuntime

class FileSystemTest extends FreeSpec with TemporaryFolder {

  "Live" - {
    "hasLocalFile" - {
      "file exists" in {
        withDirectory(dir => {
          val filename = "filename"
          createFile(dir, filename, contents = "")
          val remoteKey = RemoteKey.create(filename)
          val sources   = Sources.create(List(dir).asJava)
          val prefix    = RemoteKey.create("")
          val program   = FileSystem.hasLocalFile(sources, prefix, remoteKey)
          val result = new DefaultRuntime {}
            .unsafeRunSync(program.provide(FileSystem.Live))
            .toEither
          val expected = true
          assertResult(Right(expected))(result)
        })
      }
      "file does not exist" in {
        withDirectory(dir => {
          val filename  = "filename"
          val remoteKey = RemoteKey.create(filename)
          val sources   = Sources.create(List(dir).asJava)
          val prefix    = RemoteKey.create("")
          val program   = FileSystem.hasLocalFile(sources, prefix, remoteKey)
          val result = new DefaultRuntime {}
            .unsafeRunSync(program.provide(FileSystem.Live))
            .toEither
          val expected = false
          assertResult(Right(expected))(result)
        })
      }
    }
  }
}

File: filesystem/src/test/java/net/kemitix/thorp/filesystem/FileSystemTest.java
package net.kemitix.thorp.filesystem;

import net.kemitix.thorp.domain.RemoteKey;
import net.kemitix.thorp.domain.Sources;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.Collections;

public class FileSystemTest
        implements WithAssertions, TemporaryFolder {

    @Test
    @DisplayName("file exists")
    public void fileExists() throws IOException {
        withDirectory(dir -> {
            String filename = "filename";
            createFile(dir, filename, Collections.emptyList());
            RemoteKey remoteKey = RemoteKey.create(filename);
            Sources sources = Sources.create(Collections.singletonList(dir));
            RemoteKey prefix = RemoteKey.create("");
            boolean result = FileSystem.hasLocalFile(sources, prefix, remoteKey);
            assertThat(result).isTrue();
        });
    }
    @Test
    @DisplayName("file does not exist")
    public void fileNotExist() throws IOException {
        withDirectory(dir -> {
            String filename = "filename";
            RemoteKey remoteKey = RemoteKey.create(filename);
            Sources sources = Sources.create(Collections.singletonList(dir));
            RemoteKey prefix = RemoteKey.create("");
            boolean result = FileSystem.hasLocalFile(sources, prefix, remoteKey);
            assertThat(result).isFalse();
        });
    }
}
